rate.go

```diff
@@ -80,36 +107,76 @@ func (r *rateLimiter) Start() {
 func (r *rateLimiter) Limit(w http.ResponseWriter, req *http.Request) (<-chan bool, error) {
 	// Create return channel and initialize
 	ch := make(chan bool, 1)
+	// How long until the next request is permitted? Allow the request IFF
+	// that is less than zero (in the past).
+	var tat, newTat, now int64
 	ok := true
 	key := r.vary.Key(req)
 
-	// Get the current count and remaining seconds
-	cnt, secs, err := r.store.Incr(key, r.window)
-	// Handle the possible situations: error, begin new window, or increment current window.
-	switch {
-	case err != nil && err != ErrNoSuchKey:
-		// An unexpected error occurred
-		return nil, err
-	case err == ErrNoSuchKey || secs <= 0:
-		// Reset counter
-		if err := r.store.Reset(key, r.window); err != nil {
+	i := 0
+	for {
+		var err error
+
+		now = r.clock()
+
+		// tat refers to the theoretical arrival time that would be expected
+		// from equally spaced requests at exactly the rate limit.
+		tat, err = r.store.Get(key)
+		if err == store.ErrNoSuchKey {
+			tat = now
+		} else if err != nil {
 			return nil, err
 		}
-		cnt = 1
-		secs = int(r.window.Seconds())
-	default:
-		// If the limit is reached, deny access
-		if cnt > r.reqs {
+
+		// How long until the next request is permitted? Allow the request
+		// IFF that is less the zero (in the past).
+		if now < (tat - int64(r.delayVariationTolerance)) {
 			ok = false
+			break
+		}
+
+		if now > tat {
+			newTat = now + int64(r.emissionInterval)
+		} else {
+			newTat = tat + int64(r.emissionInterval)
+		}
+
+		var updated bool
+		if err == store.ErrNoSuchKey {
```

Comment:
This is incredibly pedantic, but mostly since there are few other problems here ;) it may be worth remembering the store's state in a boolean instead of re-using `err` here. It a call that re-uses `err` is added between here and line 125, it may result in something unexpected happening.